src/query/src/optimizer/remove_duplicate.rs

```diff
@@ -0,0 +1,117 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use common_query::DfPhysicalPlanRef;
+use datafusion::config::ConfigOptions;
+use datafusion::physical_optimizer::PhysicalOptimizerRule;
+use datafusion::physical_plan::coalesce_batches::CoalesceBatchesExec;
+use datafusion::physical_plan::repartition::RepartitionExec;
+use datafusion_common::tree_node::{Transformed, TreeNode};
+use datafusion_common::Result as DfResult;
+
+/// This is [PhysicalOptimizerRule] to remove duplicate physical plans such as two
+/// adjoining [CoalesceBatchesExec] or [RepartitionExec]. They won't have any effect
+/// if one runs right after another.
+///
+/// This rule is expected to be run in the final stage of the optimization process.
+pub struct RemoveDuplicate;
+
+impl PhysicalOptimizerRule for RemoveDuplicate {
+    fn optimize(
+        &self,
+        plan: DfPhysicalPlanRef,
+        _config: &ConfigOptions,
+    ) -> DfResult<DfPhysicalPlanRef> {
+        Self::do_optimize(plan)
+    }
+
+    fn name(&self) -> &str {
+        "RemoveDuplicateRule"
+    }
+
+    fn schema_check(&self) -> bool {
+        false
+    }
+}
+
+impl RemoveDuplicate {
+    fn do_optimize(plan: DfPhysicalPlanRef) -> DfResult<DfPhysicalPlanRef> {
+        let result = plan
+            .transform_down_mut(&mut |plan| {
+                if plan.as_any().is::<CoalesceBatchesExec>()
+                    || plan.as_any().is::<RepartitionExec>()
+                {
+                    // check child
+                    let child = plan.children()[0].clone();
+                    if child.as_any().type_id() == plan.as_any().type_id() {
+                        // remove child
+                        let grand_child = child.children()[0].clone();
```

Comment:
If I don't misunderstand, it looks like every plan node only has one child. Am I right?

Reply:
not *every* plan, but plans referenced here only have one child.